### PR TITLE
[CRT] Some NT6 fixes

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Some fixes for NT6

## Proposed changes
- Import _gcvt and _gcvt_s from wine and export _gcvt_s on Vista+
- Add parameter check to _mbscmp
- Fix parameter check for _memicmp on NT 6+

### CODE REVIEW

File: sdk/lib/crt/stdlib/gcvt.c [hunk 1]
```diff
@@ -1,32 +1,73 @@
...
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
```
Comment (JoachimHenze) [inline sdk/lib/crt/stdlib/gcvt.c:20]:
```suggestion
```
suggested unless you synced majority of that file with Wine here (which was not stated in the PRs description).

**Review by DarkFire01 — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

File: sdk/lib/crt/mem/memicmp.c [hunk 2]
```diff
@@ -8,6 +9,16 @@ int
 CDECL
 _memicmp(const void *s1, const void *s2, size_t n)
 {
+    if (NtCurrentPeb()->OSMajorVersion >= 6)
+    {
+        if (!s1 || !s2)
+        {
+            if (n)
+                MSVCRT_INVALID_PMT(NULL, EINVAL);
+            return n ? _NLSCMPERROR : 0;
```
Comment (katahiromz) [inline sdk/lib/crt/mem/memicmp.c:18]: You are checking `n` twice at here.

File: sdk/lib/crt/stdlib/gcvt.c [hunk 3]
```diff
@@ -1,32 +1,73 @@
...
-
-  /* They want Xe-YY, not X.e-YY, and XXXX instead of XXXX.  */
-  p = strchr (buf, 'e');
-  if (!p)
-    {
-      p = buf + strlen (buf);
-      /* They don't want trailing zeroes.  */
-      while (p[-1] == '0' && p > buf + 2)
-	*--p = '\0';
+    if (!buff) {
```
Comment (katahiromz) [inline sdk/lib/crt/stdlib/gcvt.c:30]:
```suggestion
    if (!buff)
    {
```

File: sdk/lib/crt/stdlib/gcvt.c [hunk 4]
```diff
@@ -1,32 +1,73 @@
...
+        *_errno() = EINVAL;
+        return NULL;
+    }
+
+    if (ndigit < 0) {
```
Comment (katahiromz) [inline sdk/lib/crt/stdlib/gcvt.c:35]:
```suggestion
    if (ndigit < 0)
    {
```

File: sdk/lib/crt/stdlib/gcvt.c [hunk 5]
```diff
@@ -1,32 +1,73 @@
...
+}
+
+/***********************************************************************
+ *              _gcvt_s  (MSVCRT.@)
+ */
+int CDECL _gcvt_s(char* buff, size_t size, double number, int digits)
+{
+    int len;
+
+    if (!buff) {
```
Comment (katahiromz) [inline sdk/lib/crt/stdlib/gcvt.c:51]:
```suggestion
    if (!buff)
    {
```

File: sdk/lib/crt/stdlib/gcvt.c [hunk 6]
```diff
@@ -1,32 +1,73 @@
...
+        *_errno() = EINVAL;
+        return EINVAL;
+    }
+
+    if (digits < 0 || digits >= size) {
```
Comment (katahiromz) [inline sdk/lib/crt/stdlib/gcvt.c:56]:
```suggestion
    if (digits < 0 || digits >= size)
    {
```

File: sdk/lib/crt/stdlib/gcvt.c [hunk 7]
```diff
@@ -1,32 +1,73 @@
...
+        if (size)
+            buff[0] = '\0';
+
+        *_errno() = ERANGE;
+        return ERANGE;
+    }
+
+    len = _scprintf("%.*g", digits, number);
+    if (len > size) {
```
Comment (katahiromz) [inline sdk/lib/crt/stdlib/gcvt.c:65]:
```suggestion
    if (len > size)
    {
```

**Review by katahiromz — COMMENTED** [review]
Minor style suggestions.

File: sdk/lib/crt/stdlib/gcvt.c (hunk 3 above)
Comment (tkreuzer) [reply sdk/lib/crt/stdlib/gcvt.c:30]: This is wine code, reformatting it doesn't make much sense.

File: sdk/lib/crt/mem/memicmp.c (hunk 2 above)
Comment (tkreuzer) [reply sdk/lib/crt/mem/memicmp.c:18]: I improved it a bit.

**Pull request merged**